### ISSUE TITLE
help resolving confusion regarding CopyGenerator

### ISSUE THREAD

**Issue by wasiahmad** [OpenNMT/OpenNMT-py#850]
In the CopyGenerator class, the input_size is the `rnn_size`. Then, why do we check the shapes as follows.

```
batch_by_tlen, _ = hidden.size()
batch_by_tlen_, slen = attn.size()
slen_, batch, cvocab = src_map.size()
aeq(batch_by_tlen, batch_by_tlen_)
aeq(slen, slen_)
```

As per documentation arguments of CopyGenerator are:

```
Args:
     hidden (`FloatTensor`): hidden outputs `[batch*tlen, input_size]`
     attn (`FloatTensor`): attn for each `[batch*tlen, input_size]`
     src_map (`FloatTensor`):
             A sparse indicator matrix mapping each source word to
             its index in the "extended" vocab containing.
             `[src_len, batch, extra_words]`
```

If `attn` is of shape: `[batch*tlen, input_size]`, then what is the meaning of the line: `aeq(slen, slen_)` in the first code snippet? As per my understanding, `attn` is of shape: `[batch*tlen, src_len]`, but in that case how can you define the followings in the constructor of the CopyGenerator class?

```
self.linear = nn.Linear(input_size, len(tgt_dict))
self.linear_copy = nn.Linear(input_size, 1)
```

[**RESOLVED**, please ignore the above content]

---------------

Also, I also need help to understand the following two functions.

[Function1](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/inputters/text_dataset.py#L232): 
```
def make_src(data, vocab):
    """ ? """
    src_size = max([t.size(0) for t in data])
    src_vocab_size = max([t.max() for t in data]) + 1
    alignment = torch.zeros(src_size, len(data), src_vocab_size)
    for i, sent in enumerate(data):
        for j, t in enumerate(sent):
            alignment[j, i, t] = 1
    return alignment
```

Why in the second line, we compute the `src_vocab_size` in that way? All the words in the source vocabulary may not appear in one batch, so what actually `src_vocab_size` means here?

[Function2](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/inputters/text_dataset.py#L246):

```
def make_tgt(data, vocab):
    """ ? """
    tgt_size = max([t.size(0) for t in data])
    alignment = torch.zeros(tgt_size, len(data)).long()
    for i, sent in enumerate(data):
        alignment[:sent.size(0), i] = sent
    return alignment
```

I assume the `make_tgt` function computes the alignment which is used while computing loss by this [function](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/copy_generator.py#L178). But the function only converts the `data` into tensor. No alignment is actually done. 

I am trying to use the CopyGenerator module from OpenNMT in other place and thus having difficulty to understand the implementation details. Answer to the questions and also if there is any good documentation on it, pointing to that doc. will be appreciated.

@vince62s can you help me? I saw you were answering other's questions, so I am tagging you. Thanks in advance.

**Comment by wasiahmad** [comment]
I cracked the underlying idea, hence closing the issue. Thanks!!

**Comment by XingLuxi** [comment]
@wasiahmad would you like to explain your understanding about `src_vocab_size ` in the Function 1 ?
